### DISCUSSION TITLE
App Directory - Route params not available in Layout when navigating from a parent page (client side navigation)

### DISCUSSION THREAD

**Discussion by nicolaslair** [vercel/next.js#45851] [Help]
Hi there,

**UPDATED - As per my answer below:**

I am building a new app using the new 'App' directory / structure. 

In the 'app/projects/[id]' route (folder), I have included a layout.tsx which is used by many sub-routes. This layout is basically getting some the URL params using params.id:

```
export default function Layout({
  children,
  params
}: {
  children: React.ReactNode,
  params: { uid: string }
  
}) {
  const project:IProject =  projects.find((p) => p.uid === params.id) as IProject;
  return (
    <div>
      <ProjectProvider data={project}>
        <Title label={`${project?.title} - ${params.id}`} />
        {children}
      </ProjectProvider>
    </div>
  )
}
```

When I navigate directly to a child page - I can see the proper title - {id} being displayed displayed in the sub route. You can check it online: https://o-gen-web-a8fw35wut-above-the-clouds.vercel.app/projects/3/clip

When I navigate to a parent page, and then click on a link to navigate to the same child page (/projects/3/clip) - so client's side navigation using next/link - the params.id is undefined and so the title / id are not displayed on the page - Check it online by clicking on any link on the following page: https://o-gen-web-a8fw35wut-above-the-clouds.vercel.app/projects/list

Everything is working locally - but not when built / deployed.

Any idea why this is happening? Any help / feedback is more than welcome ;)

Thanks in advance, Nico

**Comment by ciruz** [comment]
I understand your problem but I don't see any code to get an idea what's happening with the rendering.

But normally, you define the Provider (state/context) within the ``RootLayout`` (``app/layout.tsx``).

Check this out => https://beta.nextjs.org/docs/rendering/server-and-client-components#rendering-third-party-context-providers-in-server-components

Did you try to define it in the ``RootLayout``, does it make a difference for you?

**Reply by nicolaslair** [reply]
Thanks a lot @ciruz - the problem is actually coming from the route params not being available when navigating from a client-side link.
Client on any link from this list: https://o-gen-web-lvao87ukb-above-the-clouds.vercel.app/projects/list - the title is "UNDEFINED - %5BID%5D" - If you refresh the page it's properly displayed. 

```
export default async function Layout({ children, params }: PageProps) {

  const project:IProject =  projects.find((p) => p.uid === params.id) as IProject;

  // Bug here - params.id undefined!
  return (
    <div>
      <ProjectProvider data={project}>
        <Title label={`${project?.title} - ${params.id}`} />
            <BackButton />
```

It seems to me like a bug in nextjs - what do you think?

**Comment by Hyuga-Tsukui** [comment]
Same thing is happening to me.

When debugging the `next build`, it is obvious that some strange id's are flowing in.


Decoding `%5Bid%5D `  is an `[id]`.

![image](https://user-images.githubusercontent.com/30320033/219873023-6650d88a-b433-401c-919a-386c2d06294d.png)

When I tried skipping this unusual value, the other pages were successfully generated.

**Reply by Hyuga-Tsukui** [reply]
I was able to reproduce this event and have created [a bug report](https://github.com/vercel/next.js/issues/46098).
